### PR TITLE
IBX-6773: Bookmarks for non accessible contents cause exception

### PR DESCRIPTION
**Pull request by vidarl**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-6773](https://issues.ibexa.co/browse/IBX-6773)
| **Type**                                   | bug
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | yes

If user bookmark some location which he later looses access too, then the bookmark list in admin-ui fails with an exception.
Simply fixing `BookmarkService::loadBookmarks()` would be easy. The problem is to implement `countUserBookmarks()` in persistence layer and having it taking into account user permissions so that BC would be kept.

Talked with Adam on how to solve this without breaking BC and he suggested implementing it using filtering

The Bookmark filter will only work with location filtering (`LocationService::find()`), not with content (`ContentService::find()`)

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: eZ/Publish/Core/Repository/BookmarkService.php [hunk 1]
```diff
@@ -97,16 +100,22 @@ public function loadBookmarks(int $offset = 0, int $limit = 25): BookmarkList
...
-            $bookmarks = $this->bookmarkHandler->loadUserBookmarks($currentUserId, $offset, $limit);
-
-            $list->items = array_map(function (Bookmark $bookmark) {
-                return $this->repository->getLocationService()->loadLocation($bookmark->locationId);
-            }, $bookmarks);
+        $filter = new Filter();
+        try {
+            $filter
+                ->withCriterion(new Criterion\Bookmark($currentUserId))
+                ->withSortClause(new SortClause\DatePublished(Query::SORT_DESC))
```
Comment (vidarl) [inline eZ/Publish/Core/Repository/BookmarkService.php]: Actually, original implementation has:
```
SELECT id, name, user_id, node_id FROM ezcontentbrowsebookmark WHERE user_id = :user_id ORDER BY id DESC LIMIT 10 {":user_id":14} []
```

So in order to maintain BC in regards to order of the results, we also need to implement a new `SortClause\BookmarkId`.
Should I do that?

Comment (vidarl) [reply eZ/Publish/Core/Repository/BookmarkService.php]: We had test (`BookmarkServiceTest::testLoadBookmarks`) that specifically tested order of bookmarks. So implemented `SortClause\BookmarkId` in order to maintain BC

File: eZ/Publish/API/Repository/Values/Content/Query/Criterion/IsBookmarked.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Values\Content\Query\Criterion;
```
Comment (ViniTou) [inline eZ/Publish/API/Repository/Values/Content/Query/Criterion/IsBookmarked.php]: All new classes should be in Ibexa namespace

**Review by ViniTou — CHANGES_REQUESTED** [review]

File: eZ/Publish/Core/Persistence/Legacy/Tests/Filter/CriterionQueryBuilder/Location/BookmarkQueryBuilderTest.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+namespace eZ\Publish\Core\Persistence\Legacy\Tests\Filter\CriterionQueryBuilder\Location;
+
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\Core\Persistence\Legacy\Filter\CriterionQueryBuilder\Location\BookmarkQueryBuilder;
+use eZ\Publish\Core\Persistence\Legacy\Tests\Filter\BaseCriterionVisitorQueryBuilderTestCase;
+
+/**
+ * @covers \eZ\Publish\Core\Persistence\Legacy\Filter\CriterionQueryBuilder\Location\BookmarkQueryBuilder::buildQueryConstraint
+ * @covers \eZ\Publish\Core\Persistence\Legacy\Filter\CriterionQueryBuilder\Location\BookmarkQueryBuilder::accepts
+ */
```
Comment (Steveb-p) [inline eZ/Publish/Core/Persistence/Legacy/Tests/Filter/CriterionQueryBuilder/Location/BookmarkQueryBuilderTest.php]: `@covers` annotation should specify a class only (not the method), or not be used at all.

File: eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php [hunk 4]
```diff
@@ -84,6 +86,8 @@ public function loadUserBookmarks(int $userId, int $offset = 0, int $limit = -1)
     }
 
     /**
+     * @deprecated Please use LocationService::count() and Criterion\IsBookmarked instead.
```
Comment (Steveb-p) [inline eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php]: We will need to adjust the deprecation messages to use the required structure.

File: eZ/Publish/Core/Persistence/Legacy/Bookmark/Gateway/ExceptionConversion.php [hunk 5]
```diff
@@ -65,6 +74,13 @@ public function loadUserBookmarks(int $userId, int $offset = 0, int $limit = -1)
         }
     }
 
+    /**
+     * @deprecated Please use LocationService::count() and Criterion\IsBookmarked instead.
+     *
+     * @param int $userId
+     *
+     * @return int
```
Comment (Steveb-p) [inline eZ/Publish/Core/Persistence/Legacy/Bookmark/Gateway/ExceptionConversion.php]: Annotations that provide the same information as the actual PHP declarations should be omitted.
```suggestion
```

File: eZ/Publish/Core/Persistence/Legacy/Bookmark/Gateway/ExceptionConversion.php [hunk 6]
```diff
@@ -56,6 +56,15 @@ public function loadBookmarkDataByUserIdAndLocationId(int $userId, array $locati
         }
     }
 
+    /**
+     * @deprecated Please use LocationService::find() and Criterion\IsBookmarked instead.
+     *
+     * @param int $userId
+     * @param int $offset
+     * @param int $limit
```
Comment (Steveb-p) [inline eZ/Publish/Core/Persistence/Legacy/Bookmark/Gateway/ExceptionConversion.php]:
```suggestion
```

File: eZ/Publish/API/Repository/Values/Content/Query/SortClause/BookmarkId.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Sets sort direction on the Bookmark ID for a location query containing a IsBookmarked criterion.
+ */
+class BookmarkId extends SortClause implements FilteringSortClause
+{
+    /**
+     * Constructs a new BookmarkId SortClause.
+     *
+     * @param string $sortDirection
+     */
```
Comment (Steveb-p) [inline eZ/Publish/API/Repository/Values/Content/Query/SortClause/BookmarkId.php]:
```suggestion
```

File: eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php (hunk 4 above)
Comment (vidarl) [reply eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php]: @Steveb-p : Could you be more specific on how you want the deprecation messages?

Comment (vidarl) [reply eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php]: @Steveb-p  : Updated in 9c9b56e58a2405e152ce0f5a1145f0af0d67b94c . OK ?

File: eZ/Publish/API/Repository/Values/Content/Query/Criterion/IsBookmarked.php (hunk 2 above)
Comment (vidarl) [reply eZ/Publish/API/Repository/Values/Content/Query/Criterion/IsBookmarked.php]: Fixed

File: eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php (hunk 4 above)
Comment (Steveb-p) [reply eZ/Publish/Core/Persistence/Legacy/Bookmark/Handler.php]: Yup, that's perfect now :)

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 8]
```diff
@@ -1980,13 +1982,24 @@ public function testBookmarksAreSwappedAfterSwapLocation()
...
 
-        $this->assertEquals($beforeSwap->items[0]->id, $afterSwap->items[1]->id);
-        $this->assertEquals($beforeSwap->items[1]->id, $afterSwap->items[0]->id);
+        $expectedIdsAfter = array_map(static function (Location $item) use ($demoDesignLocationId, $contactUsLocationId) {
+            if ($item->id === $demoDesignLocationId) {
+                return $contactUsLocationId;
+            }
+
+            return $item->id;
+        }, $beforeSwap->items);
```
Comment (Steveb-p) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: Why is this mapping necessary? In particular, why is there a special case for `$demoDesignLocationId` + `$contactUsLocationId`?

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 9]
```diff
@@ -1980,13 +1982,24 @@ public function testBookmarksAreSwappedAfterSwapLocation()
...
+
+        $actualIdsAfter = array_map(static function (Location $item) use ($demoDesignLocationId, $contactUsLocationId) {
```
Comment (Steveb-p) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php]:
```suggestion
        $actualIdsAfter = array_map(static function (Location $item) {
```

File: eZ/Publish/Core/Repository/BookmarkService.php [hunk 10]
```diff
@@ -97,16 +100,22 @@ public function loadBookmarks(int $offset = 0, int $limit = 25): BookmarkList
...
-            }, $bookmarks);
+        $filter = new Filter();
+        try {
+            $filter
+                ->withCriterion(new Criterion\IsBookmarked($currentUserId))
+                ->withSortClause(new SortClause\BookmarkId(Query::SORT_DESC))
+                ->sliceBy($limit, $offset);
+
+            $result = $this->repository->getlocationService()->find($filter, []);
+        } catch (\eZ\Publish\API\Repository\Exceptions\BadStateException $e) {
```
Comment (Steveb-p) [inline eZ/Publish/Core/Repository/BookmarkService.php]: This class should be imported.

```suggestion
        } catch (BadStateException $e) {
```

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php (hunk 8 above)
Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: The mapping is no longer needed. In first version of PR, I did not implement `SortClause\BookmarkId`, so order of the result needed to be normalized.
I will revert the mapping.

I think the test is better when adding `$contactUsLocation`. In original test we test the edge case where two bookmarked locations are swapped... In my version of the test, only one of the locations that are swapped are bookmarked.

Edit: Fixed cut&paste error that made the second paragraph meaningless

Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: Done in 8483861fa87eaf7929084818623dace97d088349

ok?

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php (hunk 9 above)
Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: mapping moved in 8483861fa87eaf7929084818623dace97d088349, so no longer relevant

File: eZ/Publish/Core/Repository/BookmarkService.php (hunk 10 above)
Comment (vidarl) [reply eZ/Publish/Core/Repository/BookmarkService.php]: Fixed

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php (hunk 8 above)
Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: @Steveb-p @ViniTou : Could you have a second look on this please?

Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: ping @Steveb-p @ViniTou

File: eZ/Publish/Core/Repository/BookmarkService.php [hunk 11]
```diff
@@ -97,16 +101,22 @@ public function loadBookmarks(int $offset = 0, int $limit = 25): BookmarkList
...
+        try {
+            $filter
+                ->withCriterion(new Criterion\IsBookmarked($currentUserId))
+                ->withSortClause(new SortClause\BookmarkId(Query::SORT_DESC))
+                ->sliceBy($limit, $offset);
+
+            $result = $this->repository->getlocationService()->find($filter, []);
+        } catch (BadStateException $e) {
+            return new BookmarkList();
         }
```
Comment (Steveb-p) [inline eZ/Publish/Core/Repository/BookmarkService.php]: This exception should probably be logged. We had plenty of situations in the past where people asked why something isn't working as it should, while we "swallowed" the exception.

Comment (vidarl) [reply eZ/Publish/Core/Repository/BookmarkService.php]: Added logging as suggested. Please have a new look @Steveb-p

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php (hunk 8 above)
Comment (vidarl) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: ping @Steveb-p @ViniTou

File: eZ/Publish/Core/Repository/BookmarkService.php [hunk 12]
```diff
@@ -97,16 +104,24 @@ public function loadBookmarks(int $offset = 0, int $limit = 25): BookmarkList
...
+                ->withCriterion(new Criterion\IsBookmarked($currentUserId))
+                ->withSortClause(new SortClause\BookmarkId(Query::SORT_DESC))
+                ->sliceBy($limit, $offset);
 
-            $list->items = array_map(function (Bookmark $bookmark) {
-                return $this->repository->getLocationService()->loadLocation($bookmark->locationId);
-            }, $bookmarks);
+            $result = $this->repository->getlocationService()->find($filter, []);
+        } catch (BadStateException $e) {
+            $this->logger->debug($e);
```
Comment (Steveb-p) [inline eZ/Publish/Core/Repository/BookmarkService.php]:
```suggestion
            $this->logger->debug($e, [
                'exception' => $e,
            ]);
```

**Review by Steveb-p — APPROVED** [review]

File: eZ/Publish/Core/Repository/Tests/Service/Mock/BookmarkTest.php [hunk 13]
```diff
@@ -219,28 +220,13 @@ public function testLoadBookmarks()
...
+        $locationList = new LocationList(['totalCount' => $expectedTotalCount, 'locations' => $expectedItems]);
 
         $locationServiceMock = $this->createMock(LocationService::class);
         $locationServiceMock
-            ->expects($this->exactly($expectedTotalCount))
-            ->method('loadLocation')
-            ->willReturnCallback(function ($locationId) {
-                return $this->createLocation($locationId);
-            });
+            ->expects($this->once())
```
Comment (konradoboza) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/BookmarkTest.php]:
```suggestion
            ->expects(self::once())
```

File: src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+namespace Ibexa\Contracts\Core\Repository\Values\Content\Query\Criterion;
+
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion\Operator;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion\Operator\Specifications;
+use eZ\Publish\SPI\Repository\Values\Filter\FilteringCriterion;
+
+/**
+ * A criterion that matches locations of bookmarks for a given user id.
+ *
```
Comment (konradoboza) [inline src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php:18]:
```suggestion
```

File: src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Supported operators:
+ * - EQ: matches against a unique user id
+ */
+class IsBookmarked extends Criterion implements FilteringCriterion
```
Comment (konradoboza) [inline src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php]:
```suggestion
final class IsBookmarked extends Criterion implements FilteringCriterion
```

File: src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+{
+    /**
+     * Creates a new IsBookmarked criterion.
+     *
+     * @param int $value UserID for which bookmarked locations must be matched against
```
Comment (konradoboza) [inline src/contracts/Repository/Values/Content/Query/Criterion/IsBookmarked.php]:
```suggestion
     * @param int $value user id for which bookmarked locations must be matched against
```

File: src/contracts/Repository/Values/Content/Query/SortClause/BookmarkId.php [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+declare(strict_types=1);
+
+namespace Ibexa\Contracts\Core\Repository\Values\Content\Query\SortClause;
+
+use eZ\Publish\API\Repository\Values\Content\Query;
+use eZ\Publish\API\Repository\Values\Content\Query\SortClause;
+use eZ\Publish\SPI\Repository\Values\Filter\FilteringSortClause;
+
+/**
+ * Sets sort direction on the Bookmark ID for a location query containing a IsBookmarked criterion.
```
Comment (konradoboza) [inline src/contracts/Repository/Values/Content/Query/SortClause/BookmarkId.php]:
```suggestion
 * Sets sort direction on the bookmark id for a location query containing IsBookmarked criterion.
```

File: src/contracts/Repository/Values/Content/Query/SortClause/BookmarkId.php [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class BookmarkId extends SortClause implements FilteringSortClause
```
Comment (konradoboza) [inline src/contracts/Repository/Values/Content/Query/SortClause/BookmarkId.php]:
```suggestion
final class BookmarkId extends SortClause implements FilteringSortClause
```

File: src/lib/Persistence/Legacy/Filter/SortClauseQueryBuilder/Bookmark/IdSortClauseQueryBuilder.php [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace Ibexa\Core\Persistence\Legacy\Filter\SortClauseQueryBuilder\Bookmark;
+
+use eZ\Publish\SPI\Persistence\Filter\Doctrine\FilteringQueryBuilder;
+use eZ\Publish\SPI\Repository\Values\Filter\FilteringSortClause;
+use eZ\Publish\SPI\Repository\Values\Filter\SortClauseQueryBuilder;
+use Ibexa\Contracts\Core\Repository\Values\Content\Query\SortClause\BookmarkId;
+
+class IdSortClauseQueryBuilder implements SortClauseQueryBuilder
```
Comment (konradoboza) [inline src/lib/Persistence/Legacy/Filter/SortClauseQueryBuilder/Bookmark/IdSortClauseQueryBuilder.php]:
```suggestion
final class IdSortClauseQueryBuilder implements SortClauseQueryBuilder
```

**Review by konradoboza — COMMENTED** [review]
Please make sure the deprecated methods get removed on merging this up to `main`.

**Comment by vidarl** [comment]
Closed in favor of 4.6 port: https://github.com/ibexa/core/pull/476
